register_service/_crypto.py

```diff
@@ -0,0 +1,229 @@
+import collections
+import os
+import io
+import math
+import struct
+
+from nacl.bindings import crypto_scalarmult_base, crypto_scalarmult
+
+from cryptography.exceptions import InvalidTag
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from cryptography.hazmat.primitives.hashes import SHA512
+from cryptography.hazmat.primitives.hmac import HMAC
+from cryptography.hazmat.primitives.kdf.hkdf import HKDF
+
+# NOTE: All assert's here are logical asserts and do NOT depend on input data. I.e. they only explicate the contract.
+#       Error handling on the other hand uses NoiseError and other crypto exceptions.
+
+backend = default_backend()
+
+
+SUITE_NAME = b"Noise255/AES256-GCM\0\0\0\0\0"
+H_LEN = 64
+CV_LEN_BYTE = 48
+SYMM_KEY_LEN_BYTE = 32
+NONCE_LEN_BYTE = 12
+MAC_BYTES = 16
+HEADER_CIPHER_TEXT_LEN_BYTE = 48
+PADDING_LEN_BYTES = 4
+ASYM_KEY_SIZE_BYTE = 32
+
+
+DecryptedNoiseBox = collections.namedtuple('DecryptedNoiseBox', 'sender_pubkey contents')
+
+
+class KeyPair:
+    def __init__(self, private_key=None, public_key=None):
+        if not private_key:
+            private_key = os.urandom(ASYM_KEY_SIZE_BYTE)
+        elif isinstance(private_key, str):
+            private_key = decode_key(private_key)
+        private_key_length = len(private_key)
+        if private_key_length != ASYM_KEY_SIZE_BYTE:
+            raise ValueError('(decoded) private key has wrong length: {} != {}'.format(private_key_length, ASYM_KEY_SIZE_BYTE))
+        if not public_key:
+            public_key = crypto_scalarmult_base(private_key)
+        self.private_key = private_key
+        self.public_key = public_key
+
+    def ecdh(self, other_public_key):
+        return crypto_scalarmult(self.private_key, other_public_key)
+
+
+class NoiseError(Exception):
+    pass
+
+
+def our_kdf(secret, extra_secret, info, output_len):
+    """
+    Noise Key derivation function. Outputs a byte sequence that the caller typically splits into multiple variables
+    such as a chain variable and cipher context, or two cipher contexts.
+
+    Args:
+        secret:       secret for key derivation
+        extra_secret: is used to pass a chaining variable to mix into the KDF.
+        info:         ensures that applying the KDF to the same secret values will produce independent output,
+                        provided 'info' is different.
+        output_len:   length out the output
+
+    Returns:
+        derived key of output_len bytes (BytesIO)
+    """
+
+    # XXX either I used hkdf below wrong and never managed to use it correctly, or the function we're using ain't
+    # XXX the RFC5869 HKDF. Anyway, this is compatible with the Java implementation we have.
```

Comment:
Instead of being compatible with our implementation I would suggest to implement the correct standard in both, the index server and the core.
I guess we implemented Noise according to [this version of the protocol](https://github.com/noiseprotocol/noise_spec/blob/213237c43add01dc7f1bd8dfc4fb7fcbc6917521/noise.md#key-derivation) which is obsolete.
This brings problems for an update of the core but I am afraid we should not use a protocol which differs from the current standard.
As @enkore indicated, there might only be differences in the KDF.